### PR TITLE
websocket: set the HTTP host header

### PR DESCRIPTION
**Pull request by MarcoPolo**
Closes #1829. Fixes a regression in not sending the host header. Sends the host header on WSS dials when an SNI is specified. Also adds a test to verify the host header is seen by the server.

@thibmeu wrote the original patch and I contributed to the test.

### CODE REVIEW

File: p2p/transport/websocket/websocket.go [hunk 1]
```diff
@@ -186,6 +187,12 @@ func (t *WebsocketTransport) maDial(ctx context.Context, raddr ma.Multiaddr) (ma
 			copytlsClientConf := t.tlsClientConf.Clone()
 			copytlsClientConf.ServerName = sni
 			dialer.TLSClientConfig = copytlsClientConf
+			ipAddr := wsurl.Host
+			dialer.NetDial = func(network, address string) (net.Conn, error) {
```
Comment (marten-seemann) [inline p2p/transport/websocket/websocket.go:191]: Why is this needed? We're dealing with (DNS-) resolved addresses here anyway, right?

File: p2p/transport/websocket/websocket.go [hunk 2]
```diff
@@ -186,6 +187,12 @@ func (t *WebsocketTransport) maDial(ctx context.Context, raddr ma.Multiaddr) (ma
...
+				tcpAddr, _ := net.ResolveTCPAddr(network, ipAddr)
```
Comment (marten-seemann) [inline p2p/transport/websocket/websocket.go:192]: Can we handle the error here?

File: p2p/transport/websocket/websocket_test.go [hunk 3]
```diff
@@ -218,6 +221,45 @@ func getTLSConf(t *testing.T, ip net.IP, start, end time.Time) *tls.Config {
...
+	require.NoError(t, err)
+	defer server.Close()
+
+	errChan := make(chan error, 1)
+	go func() {
+		server.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if !strings.Contains(r.Host, "example.com") {
+				errChan <- errors.New("Didn't see host header")
+			} else {
+				close(errChan)
```
Comment (marten-seemann) [inline p2p/transport/websocket/websocket_test.go:236]: `defer` the `close`?

File: p2p/transport/websocket/websocket_test.go [hunk 4]
```diff
@@ -218,6 +221,45 @@ func getTLSConf(t *testing.T, ip net.IP, start, end time.Time) *tls.Config {
...
+			}
+			w.WriteHeader(http.StatusNotFound)
+		})
+		server.TLSConfig = getTLSConf(t, net.ParseIP("127.0.0.1"), time.Now(), time.Now().Add(time.Hour))
+		server.ServeTLS(l, "", "")
+	}()
+
+	parts := strings.Split(l.Addr().String(), ":")
```
Comment (marten-seemann) [inline p2p/transport/websocket/websocket_test.go:244]: `net.SplitHostAndPort`?

File: p2p/transport/websocket/websocket.go (hunk 1 above)
Comment (MarcoPolo) [reply p2p/transport/websocket/websocket.go:193]: Added a comment

**Review by marten-seemann — APPROVED** [review]
